### PR TITLE
ENH Add a switch to override existing global variables

### PR DESCRIPTION
**Pull request by pierreglaser**
Following #214.
This PR implements a switch called `override_existing_globals`, available from `CloudPickler.__init__`, `cloudpickle.dumps` and `cloudpickle.dump`.

* If set to `True`, the globals from a function originally created in the  `__main__` module, dynamic modules or in a nested scope override the globals of the function's module at unpickling time.
* If set to `False`, the existing variables in the function's module overrides the globals of those functions at unpickling time.

### CODE REVIEW

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -577,6 +591,7 @@ def save_function_tuple(self, func):
         if hasattr(func, '__qualname__'):
             state['qualname'] = func.__qualname__
         save(state)
+        save(self.override_existing_globals)
```
Comment (suquark) [inline cloudpickle/cloudpickle.py]: Could you put this variable inside `state`? Because in the future there will could be more switches and when the argument count reaches 5 there will be no way to tell it from the older implementation.

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -278,6 +278,20 @@ def __init__(self, file, protocol=None):
         # map ids to dictionary. used to ensure that functions can share global env
         self.globals_ref = {}
 
+        # When unpickling a function created in a dynamic module or the __main__
+        # module, its set of global variables may be different from the one from
+        # the module the function belongs to, for example if the dynamic module
+        # has had its variables mutated between pickling and unpickling time
```
Comment (tomMoral) [inline cloudpickle/cloudpickle.py]: I find this misleading. I would say something like:
```suggestion
        # has had a global variable mutated between two call to pickle.dump,
        # then the two pickled objects have different value for the same variable.
```

File: cloudpickle/cloudpickle.py [hunk 3]
```diff
@@ -278,6 +278,20 @@ def __init__(self, file, protocol=None):
...
+        # of the function.
+
+        # - If set to True, the attribue override_existing_globals will
```
Comment (tomMoral) [inline cloudpickle/cloudpickle.py]:
```python
        # - If override_existing_globals=True, the global variables of the module
        # containing this function will be overridden in the process which load
        # the pickled object with the values of these variables at the pickling
        # time.
        # - If override_existing_globals=False, the global variables of the module
        # containing this function will not be overridden in the process which load
        # the pickled object. If the global variable is not declared, it will be
        # initialized with the value at pickling time.
```

File: tests/cloudpickle_test.py [hunk 4]
```diff
@@ -78,6 +78,7 @@ def setUp(self):
 class CloudPickleTest(unittest.TestCase):
 
     protocol = cloudpickle.DEFAULT_PROTOCOL
+    override_existing_globals = False
```
Comment (tomMoral) [inline tests/cloudpickle_test.py]: Why do you use a class variable here?

File: tests/cloudpickle_test.py [hunk 5]
```diff
@@ -1078,13 +1081,19 @@ def f1():
...
+            assert result_pickled_f1 == "default_value", result_pickled_f1
+        else:
+            assert result_pickled_f1 == "changed_by_f0", result_pickled_f1
         """
         for clone_func in ['local_clone', 'subprocess_pickle_echo']:
-            code = code_template.format(protocol=self.protocol,
-                                        clone_func=clone_func)
+            code = code_template.format(
+                protocol=self.protocol, clone_func=clone_func,
+                override_existing_globals=self.override_existing_globals)
```
Comment (tomMoral) [inline tests/cloudpickle_test.py]: Why don't you loop on `override_existing_globals`?

File: tests/cloudpickle_test.py [hunk 6]
```diff
@@ -1307,6 +1344,7 @@ def g(x):
 class Protocol2CloudPickleTest(CloudPickleTest):
 
     protocol = 2
+    override_existing_globals = True
```
Comment (tomMoral) [inline tests/cloudpickle_test.py]: Ok I get it now....

I find this hard to parse. I think it would be more readable and complete to loop over `[True, False]` for the two given tests.

File: tests/cloudpickle_test.py (hunk 4 above)
Comment (pierreglaser) [reply tests/cloudpickle_test.py]: To parametrize the tests. `Protocol2CloudPickleTest` uses `override_existing_globals = True`.

**Comment by pierreglaser** [comment]
@suquark `override_existing_globals` is now in `state`. 
@tomMoral I now loop over `override_existing_globals` in the test. Because of that, the diff looks messy. You can refer to [this commit](https://github.com/cloudpipe/cloudpickle/pull/216/commits/0fca5af10f206103347f7384bd5306d4055849af) for a cleaner diff.

File: cloudpickle/cloudpickle.py [hunk 7]
```diff
@@ -278,6 +278,24 @@ def __init__(self, file, protocol=None):
...
+        # Functions belonging to the same module (the __main__ or a dynamic
+        # module) share the same global variables. Using cloudpickle, they will
+        # share those global variables in the process they are unpickled in as
+        # well. However, the state of those global variables may differ in the
+        # pickled functions, for example if one global variable was mutated
+        # between two calls to pickle.dump. override_existing_globals is a
+        # switch that allows two different behaviors:
+
+        # - If override_existing_globals=True, the global variables of the
+        # module containing this function will be overridden in the process
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: "module containing the definition of the nested or dynamically defined function"

File: cloudpickle/cloudpickle.py [hunk 8]
```diff
@@ -269,7 +269,7 @@ class CloudPickler(Pickler):
 
     dispatch = Pickler.dispatch.copy()
 
-    def __init__(self, file, protocol=None):
+    def __init__(self, file, protocol=None, override_existing_globals=True):
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: In retrospect I find `override_existing_globals` a bit too verbose and redundant. May I suggest to rename this to `override_globals`.

File: tests/cloudpickle_test.py [hunk 9]
```diff
@@ -1078,129 +1080,167 @@ def f1():
...
+                    assert res_pickled_f1 == "default_value", res_pickled_f1
+                else:
+                    assert res_pickled_f1 == "changed_by_f0", res_pickled_f1
+            finally:
+                _TEST_GLOBAL_VARIABLE = orig_value
 
     def test_function_from_dynamic_module_with_globals_modifications(self):
         # This test verifies that the global variable state of a function
         # defined in a dynamic module in a child process are not reset by
         # subsequent uplickling.
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: typo: unpickling

File: tests/cloudpickle_test.py [hunk 10]
```diff
@@ -1078,129 +1080,167 @@ def f1():
...
+                # Simple sanity check on the function's output
+                assert mod.func_defined_in_dynamic_module() == "initial value"
+
+                # The function of mod is pickled two times, with two different
+                # values for the global variable GLOBAL_STATE.  Then we launch
+                # a child process that sequentially unpickles the two
+                # functions. Those unpickle functions should share the same
+                # global variables in the child process: Once the first
+                # function gets unpickled, mod is created and tracked in the
+                # child environment. This is state is preserved when unpickling
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: ... child environment. If override_globals is False, then the module state is preserved when unpickling...

File: tests/cloudpickle_test.py [hunk 11]
```diff
@@ -1078,129 +1080,167 @@ def f1():
...
+                    # override_existing_globals was True (resp. False) at
+                    # pickling time
+
+                    if {override_existing_globals}:
+                        assert func_with_modified_globals() == 'changed value'
+                    else:
+                        assert func_with_modified_globals() == 'initial value'
+
+                    # Update the value from the child process and check that
+                    # unpickling reset (resp. does not reset) our change if
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: unpickling resets

**Comment by mitar** [comment]
@suquark Does this address fully #214?

**Comment by ogrisel** [comment]
@suquark could you please provide more details? Do you have a standalone example of one such problematic recursive function?

**Comment by ogrisel** [comment]
Closing in favor of #240.
